File: examples/Delphi/Compound/ModelEvolution/Version2/dPersistence.pas
unit dPersistence;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  BoldAbstractObjectUpgraderHandle, BoldObjectUpgraderHandle,
  BoldSubscription, BoldHandle, BoldPersistenceHandle,
  BoldSystem,
  BoldPersistenceHandleDB, DB, IBDatabase,
  BoldAbstractDatabaseAdapter, BoldDatabaseAdapterIB,
  BoldAbstractPersistenceHandleDB;

type
  TdmPersistence = class(TDataModule)
    ObjectUpgrader: TBoldObjectUpgraderHandle;
    BoldPersistenceHandleDB1: TBoldPersistenceHandleDB;
    BoldDatabaseAdapterIB1: TBoldDatabaseAdapterIB;
    IBDatabase1: TIBDatabase;
    procedure ObjectUpgrader_Person_UpgradeObject(Obj: TBoldObject);
    procedure ObjectUpgrader_Order_UpgradeObject(Obj: TBoldObject);
    procedure ObjectUpgrader_Product_UpgradeObject(Obj: TBoldObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  dmPersistence: TdmPersistence;

implementation

uses dModel, ModelEvClasses, dSystemTypeInfo;

{$R *.DFM}

procedure TdmPersistence.ObjectUpgrader_Person_UpgradeObject(Obj: TBoldObject);
begin
  (Obj as TPerson).Upgrade;
end;

procedure TdmPersistence.ObjectUpgrader_Order_UpgradeObject(Obj: TBoldObject);
begin
  (Obj as TOrder).Upgrade;
end;

procedure TdmPersistence.ObjectUpgrader_Product_UpgradeObject(Obj: TBoldObject);
begin
  (Obj as TProduct).Upgrade;
end;

end.

File: examples/CBuilder/Compound/ModelEvolution/Version1/dPersistence.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "dPersistence.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BoldAbstractObjectUpgraderHandle"
#pragma link "BoldHandle"
#pragma link "BoldObjectUpgraderHandle"
#pragma link "BoldPersistenceHandle"
#pragma link "BoldPersistenceHandleBDE"
#pragma link "BoldPersistenceHandleDB"
#pragma link "BoldSubscription"
#pragma resource "*.dfm"
TdmPersistence *dmPersistence;
//---------------------------------------------------------------------------
__fastcall TdmPersistence::TdmPersistence(TComponent* Owner)
  : TDataModule(Owner)
{
}
//---------------------------------------------------------------------------
